design/service-apis.md

```diff
@@ -0,0 +1,249 @@
+Contour. While many use-cases have client traffic originating “outside” the cluster, this is not a requirement.
+
+A Gateway is 1:1 with the lifecycle of the infrastructure configuration. When a Gateway is created, the operator will
+create an instance of Contour. Gateway is the resource that triggers actions in Service APIs. Other Service API
+resources represent structured configuration that are linked together by a Gateway.
+
+A Gateway MAY contain one or more Service APIs *Route references which serve to direct traffic for a subset of traffic
+to a specific service. Note that Contour Operator is NOT responsible for managing *Route Service API resources. The
+following is an example of a Gateway that listens for HTTP traffic on TCP port 80: 
+
+```yaml
+kind: Gateway
+apiVersion: networking.x-k8s.io/v1alpha1
+metadata:
+  name: example-gateway
+spec:
+  gatewayClassName: example-gatewayclass
+  listeners:
+    - protocol: HTTP
+      port: 80
+  ...
+```
+
+The `port` or `protocol` fields are required and specify the network port/protocol pair that Envoy will listen for
+connection requests on. TLS configuration must be included in the Gateway listener when "HTTPS" or "TLS" is specified
+for `protocol`. `certificateRef` of a TLS configuration is a reference to a Secret in the operator's namespace
+containing a TLS certificate and private key. The secret MUST contain the following keys and data:
+
+- tls.crt: certificate file contents
+- tls.key: key file contents
+  
+For example:
+
+```yaml
+kind: Gateway
+...
+listeners:
+- protocol: HTTPS
+  port: 443
+  tls:
+    mode: Terminate
+    certificateRef:
+      kind: Secret
+      group: core
+      name: default-cert
+---
+kind: Secret
+apiVersion: v1
+metadata:
+  name: default-cert
+  namespace: contour-operator
+data:
+  tls.crt: <CERT>
+  tls.key: <KEY>
+```
+
+Gateway contains a `routes` field that specifies a schema for associating routes with the `listener` using Kubernetes
+selectors. A Route is a Service APIs resource capable of servicing a request and allows a user to expose a resource
+(i.e. Service) by an externally-reachable URL, proxy traffic to the resource, and terminate SSL/TLS connections. The
+following example will bind the Gateway to Routes that include label "app: example".The `port` or `protocol` fields are required when
+
+```yaml
+kind: Gateway
+apiVersion: networking.x-k8s.io/v1alpha1
+metadata:
+  name: example-gateway
+spec:
+  gatewayClassName: example-gatewayclass
+  listeners:
+    - protocol: HTTP
+      port: 80
+      routes:
+        kind: HTTPRoute
+        selector:
+          matchLabels:
+            app: example
+---
+kind: HTTPRoute
+apiVersion: networking.x-k8s.io/v1alpha1
+metadata:
+  name: example-httproute
+  labels:
+    app: example
+spec:
+  rules:
+  - forwardTo:
+    - serviceName: example-svc
+      port: 8080
+```
+
+The `addresses` field specifies one or more network addresses requested for this Gateway. If unspecified, an IP address
```

Comment:
How does the operator determine if the address is invalid or unavailable?

Reply:
@stevesloka the operator will create the Envoy service with the provided address. The Service field that is set will depend on the network publishing type specified in the `ContourConfig`. For example, if the network publishing type is "Internal", the operator will create an Envoy service of type `ClusterIP` and set `clusterIPs: 1.2.3.4`. If service creation fails with "error: address unavailable" or "error: address in use", then the operator will surface a `Gateway` status condition.

Reply:
ahh ok so you're going to surface the k8s service creation error back, that can work. I thought you were going to need to understand somehow if the address requested would work for each specific environment which seemed like a large scope.